lib/Conversion/HWToSystemC/HWToSystemC.cpp

```diff
@@ -100,32 +101,144 @@ struct ConvertHWModule : public OpConversionPattern<HWModuleOp> {
-/// values to the output block arguments.
-struct ConvertOutput : public OpConversionPattern<OutputOp> {
+/// Convert hw.instance operations to systemc.instance.decl and a
+/// systemc.instance.bind_port operation for each port in the constructor. Also
+/// insert the necessary intermediate signals and write or read their state in
+/// the update function accordingly.
+class ConvertInstance : public OpConversionPattern<InstanceOp> {
   using OpConversionPattern::OpConversionPattern;
 
+private:
+  template <typename PortTy>
+  LogicalResult
+  collectPortInfo(ValueRange ports, ArrayAttr portNames,
+                  SmallVector<ModuleType::PortInfo> &portInfo) const {
+    for (auto inPort : llvm::zip(ports, portNames)) {
+      Type ty = std::get<0>(inPort).getType();
+      ModuleType::PortInfo info;
+
+      if (ty.isa<hw::InOutType>())
+        return failure();
+
+      info.type = PortTy::get(ty);
+      info.name = std::get<1>(inPort).cast<StringAttr>();
+      portInfo.push_back(info);
+    }
+
+    return success();
+  }
+
+public:
   LogicalResult
-  matchAndRewrite(OutputOp outputOp, OpAdaptor adaptor,
+  matchAndRewrite(InstanceOp instanceOp, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
+    // Make sure the parent is already converted such that we already have a
+    // constructor and update function to insert operations into.
+    auto scModule = instanceOp->getParentOfType<SCModuleOp>();
+    if (!scModule)
+      return rewriter.notifyMatchFailure(instanceOp,
+                                         "parent was not an SCModuleOp");
+
+    // Get the builders for the different places to insert operations.
+    auto ctor = scModule.getOrCreateCtor();
+    OpBuilder stateBuilder(ctor);
+    OpBuilder initBuilder = OpBuilder::atBlockEnd(ctor.getBodyBlock());
+
+    // Collect the port types and names of the instantiated module and convert
+    // them to appropriate systemc types.
+    SmallVector<ModuleType::PortInfo> portInfo;
+    if (failed(collectPortInfo<InputType>(adaptor.getInputs(),
+                                          adaptor.getArgNames(), portInfo)) ||
+        failed(collectPortInfo<OutputType>(instanceOp->getResults(),
+                                           adaptor.getResultNames(), portInfo)))
+      return instanceOp->emitOpError("inout ports not supported");
+
+    Location loc = instanceOp->getLoc();
+    auto instanceName = instanceOp.getInstanceNameAttr();
+    auto instModuleName = instanceOp.getModuleNameAttr();
+
+    // Declare the instance.
+    auto instDecl = stateBuilder.create<InstanceDeclOp>(
+        loc, instanceName, instModuleName, portInfo);
+
+    // Bind the input ports.
+    for (size_t i = 0, numInputs = adaptor.getInputs().size(); i < numInputs;
+         ++i) {
+      Value input = adaptor.getInputs()[i];
+      APInt portId = APInt(32, i);
+      std::string signalName =
```

Comment:
Does this need to create a `std::string` here or can this instead pass a `Twine` into the builder? Ditto for later.

Reply:
Passing the `Twine` directly does not work because there is no implicit conversion to a `StringRef`. It's probably best to directly pass a `StringAttr` which accepts a `Twine` as argument.